discojs/discojs-core/src/aggregator/base.ts

```diff
@@ -0,0 +1,239 @@
+   * peer/client within the network, whom we are communicating with during this aggregation
+   * round.
+   * @param nodeId The node to be added
+   */
+  registerNode (nodeId: client.NodeID): boolean {
+    if (!this.nodes.has(nodeId)) {
+      this._nodes = this._nodes.add(nodeId)
+      return true
+    }
+    return false
+  }
+
+  /**
+   * Overwrites the current set of "nodes" with the given one. A node represents
+   * an active neighbor peer/client within the network, whom we are communicating with
+   * during this aggregation round.
+   * @param nodeIds The new set of nodes
+   */
+  setNodes (nodeIds: Set<client.NodeID>): void {
+    this._nodes = nodeIds
+  }
+
+  resetNodes (): void {
+    this._nodes = Set()
+  }
+
+  /**
+   * Sets the aggregator's round. To be used whenever the aggregator is out of sync
+   * with the network's round.
+   * @param round The new round
+   */
+  setRound (round: number): void {
+    if (round > this.round) {
+      this._round = round
+    }
+  }
+
+  /**
+   * Emits the event containing the aggregation result, which allows the result
+   * promise to resolve and for the next aggregation round to take place.
+   * @param aggregated The aggregation result
+   */
+  protected emit (aggregated: T): void {
+    this.eventEmitter.emit('aggregation', aggregated)
+  }
+
+  /**
+   * Resets the aggregator's step and prepares it for the next aggregation round.
+   */
+  public nextRound (): void {
+    if (++this._communicationRound === this.communicationRounds) {
+      this._communicationRound = 0
+      this._round++
+      this.contributions = Map()
+    }
+    this.result = this.makeResult()
+    this.informant?.update()
+  }
+
+  private async makeResult (): Promise<T> {
+    return await new Promise((resolve) => {
+      this.eventEmitter.once('aggregation', (w) => {
+        resolve(w)
+      })
+    })
+  }
+
+  /**
+   * The aggregation result can be awaited upon in an asynchronous fashion, to allow
+   * for the receipt of contributions while performing other tasks. This function
+   * gives access to the current aggregation result's promise, which will eventually
+   * resolve and contain the result of the very next aggregation step, at the
+   * time of the function call.
+   * @returns The promise containing the aggregation result
+   */
+  async receiveResult (): Promise<T> {
+    return await this.result
+  }
+
+  /**
+   * Constructs the payload sent to other nodes as contribution.
+   * @param base Object from which the payload is computed
+   */
+  abstract makePayloads (base: T): Map<client.NodeID, T>
+
+  abstract isFull (): boolean
+
+  /**
+   * The set of node ids, representing our neighbors within the network.
+   */
+  get nodes (): Set<client.NodeID> {
+    return this._nodes
+  }
+
+  /**
+   * The aggregator's round.
+   */
+  get round (): number {
+    return this._round
+  }
+
+  /**
+   * The aggregator's current size, defined by its amount of contributions.
```

Comment:
good. maybe can say this is the size of the aggregation group, even if a bit redundant

Reply:
this docstring is outdated so I updated it in #580 (and it's clearer too)